### PR TITLE
PCGrad gradient surgery: resolve τ_y/τ_z vs surface pressure gradient conflict

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

We train on 6 physical fields simultaneously (cp, τ_x, τ_y, τ_z, volume_pressure, wall_shear). The τ_y/τ_z fields are confirmed training laggards (τ_z r_i=0.01123, wall shear is the hardest channel). This suggests **gradient conflict** between fields: when gradients from τ_y/τ_z point in an opposite direction to gradients from surface pressure (cp), the optimizer moves in a direction that's suboptimal for both.

**PCGrad** (Yu et al., 2020) addresses multi-task gradient conflict by projecting conflicting task gradients onto each other's normal planes. For tasks i and j where gradient conflict exists (cos similarity < 0), the gradient for task i is replaced by its projection onto the plane orthogonal to task j's gradient. This resolves interference without requiring per-task loss weight tuning (which we've already exhausted).

PCGrad has never been tested on this problem. Given the confirmed multi-field gradient conflict in DrivAerML (τ_y/τ_z vs surface pressure), this is a targeted, principled intervention.

## Instructions

Implement PCGrad as a gradient surgery wrapper. The key operation:

```python
def pcgrad_surgery(loss_list, optimizer):
    """
    PCGrad: Project conflicting gradients.
    loss_list: list of per-task scalar losses [L_cp, L_tauy, L_tauz, L_vol, ...]
    """
    grad_list = []
    # Compute per-task gradients
    for loss in loss_list:
        optimizer.zero_grad()
        loss.backward(retain_graph=True)
        grad_list.append({name: p.grad.clone() for name, p in model.named_parameters() if p.grad is not None})
    
    # For each task, project away conflicting gradient components
    projected_grads = []
    for i, grads_i in enumerate(grad_list):
        projected = {k: v.clone() for k, v in grads_i.items()}
        for j, grads_j in enumerate(grad_list):
            if i == j:
                continue
            for k in projected:
                if k not in grads_j:
                    continue
                g_i = projected[k].flatten()
                g_j = grads_j[k].flatten()
                cos_sim = torch.dot(g_i, g_j) / (g_i.norm() * g_j.norm() + 1e-8)
                if cos_sim < 0:  # Conflicting gradients
                    # Project g_i to remove component along g_j
                    projected[k] -= (cos_sim * g_j.view_as(projected[k]))
        projected_grads.append(projected)
    
    # Sum projected gradients and apply
    optimizer.zero_grad()
    for name, param in model.named_parameters():
        if param.requires_grad:
            combined = sum(pg.get(name, torch.zeros_like(param)) for pg in projected_grads)
            param.grad = combined
    optimizer.step()
```

Apply PCGrad across these task groups: [surface_pressure_loss, tau_y_loss, tau_z_loss, volume_loss]. Keep τ_y weight=1.5, τ_z weight=2.0, surface weight=2.0 (existing optimal weights).

**NOTE:** PCGrad requires computing gradients per-task before summing. This will increase memory usage. If OOM, reduce batch size to 2 and note it.

Use the SOTA 4-ep base command:

```bash
cd /workspace/senpai/target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent tanjiro --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 4 \
  --vol-points-schedule "0:16384:1:32768:2:49152:3:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --wandb-group pcgrad-r18 --wandb-name tanjiro/pcgrad-4tasks
```

**Kill gates (4-ep screen):**
- EP1: val_abupt < 30%
- EP2: val_abupt < 16%
- EP3: val_abupt < 8%
- EP4: val_abupt ≤ 6.5985% to beat SOTA

Report EP1–EP4 val_abupt, per-channel errors (especially τ_y/τ_z vs baseline), and W&B run ID.
Also report estimated % of gradient conflicts detected (cos_sim < 0 rate) if you log it.

## Baseline

| Metric | Value | PR |
|--------|-------|-----|
| val_abupt (single-model SOTA) | **6.5985%** | #592 (run `4k25s25e`) |
| test_abupt | 7.9915% | #592 |
| test_vol_pressure | 11.933% | #592 |
| τ_z (training laggard) | r_i=0.01123 confirmed Round 17 | #758 |
| val_abupt (ensemble SOTA) | 6.1751% | #612 |

W&B project: `wandb-applied-ai-team/senpai-v1-drivaerml-ddp8`

**Reference:** Yu et al., "Gradient Surgery for Multi-Task Learning" (NeurIPS 2020): https://arxiv.org/abs/2001.06782

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT tanjiro:

## Launched

Implementation committed (e2e48b6) and 4-epoch DDP run started.

- W&B: https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/7v0rlsps
- W&B group: `pcgrad-r18`, run name `tanjiro/pcgrad-4tasks`
- All 8 GPUs at 90%+ utilisation, ~80–96 GB / 80 GB H100 memory in steady state
- PCGrad startup banner: `tasks=('surface', 'tau_y', 'tau_z', 'vp'); effective coeffs sp/taux=0.500 tau_y=0.750 tau_z=1.000 vp=1.000` (= surface_w/4 surface scaling so the no-conflict sum exactly matches the existing weighted_channel_mse baseline)

### Implementation notes

- 4 task groups: `[surface=sp+tau_x, tau_y, tau_z, vp]`. tau_x is folded into the surface task because the PR specifies 4 task groups but the model still needs gradient signal on tau_x; the surface task weight is divided by SURFACE_NUM_CHANNELS=4 so `sum(L_i)` reproduces the baseline.
- DDP-correct: per-task gradients via `torch.autograd.grad` (does NOT trigger DDP all-reduce hooks), PCGrad surgery on each rank, then a single `dist.all_reduce(SUM) + div_(world_size)` per param before `optimizer.step`.
- Conflict masking is `(dot < 0).to(dot.dtype)` with unconditional projection correction — avoids per-pair host-device sync (one final `.item()` for metrics).
- Random projection ordering seeded by `pcgrad_seed + global_step` (identical across ranks for determinism).
- Logging: `pcgrad/{conflict_count, total_pairs, conflict_rate, combined_grad_norm, grad_norm_pre_*, grad_norm_post_*}`.

### Timing concern

PCGrad runs K=4 backward passes per step. Wall-clock budget is 360 min (`SENPAI_TIMEOUT_MINUTES=360`); SOTA L=5 stack baseline is ~235 min for 4 epochs. With ~2× total slowdown I expect ~3 epochs to fit in 360 min — will report whatever completes. EP1 ETA TBC after first epoch lands.

Will follow up once EP1 val_abupt is in.

**Comment by morganmcg1** [comment]
STUDENT tanjiro:

## Mid-EP1 status (step ~1428 / 10864 = 13%)

- W&B: https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/7v0rlsps
- Wall clock: 28 min elapsed, EP1 ETA ~220 min total → EP1 expected ~3:30 from launch.
- Per-iter: ~1.24 s/it (vs ~0.6 s/it baseline → ~2× slowdown as expected with K=4 backward passes)
- 360-min budget will fit ~1.6 epochs. Will report whatever validates.

## Key observation: gradient conflict is rare

PCGrad metrics over the first 1454 logged steps:

| Stat | Value |
|------|-------|
| mean conflict_rate | **0.088** (8.8%) |
| median conflict_rate | 0.000 |
| frac steps with any conflict | 28.2% |
| frac steps with conflict_rate ≥ 0.5 | 6.3% |
| conflict_count distribution (per step, out of 12 pairs) | 0: 71.8%, 2: 12.5%, 4: 7.4%, 6: 4.1%, 8+: 4.0% |

So **the premise of the hypothesis** — that τ_y/τ_z gradients are systematically conflicting with surface pressure — **is largely false on this run**. ~72% of steps have zero conflict among any of the 12 task-pair comparisons; the optimizer is essentially seeing the baseline gradient sum on most steps. PCGrad is paying 2× compute cost to fire on only ~28% of steps.

If val_abupt by EP-end isn't materially better than baseline, this is the likely reason — "resolve gradient conflict" only helps when conflict is actually present and the unprojected sum is consistently moving away from a Pareto-improving direction. That doesn't appear to be the regime we're in.

Will follow up with EP1 val_abupt + per-channel τ_y/τ_z when validation lands.

**Comment by morganmcg1** [comment]
STUDENT tanjiro:

## Restart with batch_size=2 (after EP2 OOM)

Previous run (`7v0rlsps`) **completed EP1** but OOM'd at start of EP2 — the vol-points curriculum jumps from 16384 → 32768, and PCGrad's 4-task gradient storage + 4 backward passes pushed GPU 4 to 94.78GB / 94.97GB before the new activations could allocate.

Per the PR's "If OOM, reduce batch size to 2 and note it" instruction, restarting with `--batch-size 2` (everything else identical):

- Old run (crashed): https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/7v0rlsps
- New run (running): https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/o73mnz13

### Headline EP1 result from old run (bs=4)

| Metric | Value | Kill gate | Pass? |
|---|---|---|---|
| **val_abupt** | **25.859%** | <30% | ✅ |
| surface_pressure_rel_l2 | 19.491% | — | — |
| volume_pressure_rel_l2 | 16.096% | — | — |
| wall_shear_x_rel_l2 | 24.961% | — | — |
| wall_shear_y_rel_l2 | 34.103% | — | — |
| wall_shear_z_rel_l2 | 34.646% | — | — |

EP1 wall-clock = 13174s = **220 min** (vs ~30 min/epoch on the L=5 SOTA baseline), i.e. **~7× slowdown** — even at bs=2 we will not fit 4 epochs in the 360-min budget. Expect ~1.5–2 epochs from the bs=2 run.

### PCGrad health over the FULL 10,865 logged steps of EP1

- mean conflict_rate: **3.69%** (median 0.0%, p90 0.167, p99 0.500, max 0.833)
- frac steps with **any** conflict: **13.4%** (vs 28.2% over the first 1454 steps mentioned earlier — conflict rate **decreased** as training progressed)
- frac steps with rate ≥0.5 (≥6 of 12 task-pairs in conflict): 1.85%
- conflict_count distribution (per step out of 12 pairs): `0: 86.6%`, `2: 6.9%`, `4: 3.2%`, `6: 1.5%`, `8: 0.2%`, `10: 0.04%`

**Premise check.** The hypothesis stated τ_y/τ_z gradients are systematically conflicting with surface pressure gradients. Across 10k+ EP1 steps and 12 task-pair comparisons per step, **86.6% of steps had zero conflict among any pair**. PCGrad reduces to the standard sum-of-tasks gradient on the vast majority of steps; on the ~13% where it does fire, the projection touches at most 2 of 12 pairs in 6.9% of steps and is rarer still beyond that. The optimizer is essentially seeing the baseline weighted-MSE gradient most of the time, and the EP1 val_abupt = 25.86% is consistent with the baseline trajectory at EP1 (consistent with kill-gate <30%).

Will follow up with a final Results comment once the bs=2 run reaches a validation epoch.

**Comment by morganmcg1** [comment]
ADVISOR: Merge conflict against `tay`. Please rebase onto latest `tay` and force-push.

After rebase, continue your bs=2 PCGrad run (W&B: o73mnz13) and post the EP1-EP4 metrics to this PR when complete (kill gates: EP1 <30%, EP2 <16%, EP3 <8%, EP4 <=6.6%).

Note for analysis: the EP1 conflict-rate finding (only 3.69% of steps show gradient conflict across 12 task pairs) is itself an interesting scientific result — please ensure this gets recorded clearly in your final results comment regardless of the val_abupt outcome. If PCGrad fails to beat baseline, the conflict-rate stats tell us multi-task conflict is rare on this problem and we should redirect attention elsewhere.

Schedule:
1. `git fetch origin && git rebase origin/tay`
2. Resolve any conflicts (likely just train.py imports / arg additions)
3. `git push --force-with-lease`
4. Resume bs=2 run (or restart if checkpoint is incompatible after rebase)
5. Post results when EP4 completes.

**Comment by morganmcg1** [comment]
STUDENT tanjiro:

## Rebase complete

Rebased onto latest `tay` (now at `2e35ee3`) and force-pushed. New head is `a9dae04`.

- Conflict was a single help-text adjacency in `train.py` between PR #823's `use_surf_to_vol_xattn` entry and PCGrad's `use_pcgrad`/`pcgrad_seed` entries; kept both.
- All Config defaults preserved (`use_surf_to_vol_xattn=False`), so the existing PCGrad command line still produces the same architecture as the original launch.
- `python -c "import ast; ast.parse(open('train.py').read())"` passes.

### New SOTA context

Latest `tay` includes PR #823 (nezuko surf-vol-xattn), which set a new single-model SOTA at **val_abupt = 6.4407%** (vs the 6.5985% in this PR's baseline table). Our EP4 kill gate of <=6.5985% is now to beat the OLD SOTA, not the new one. The bs=2 PCGrad run will, in any case, only complete ~1.5 epochs in the 360-min budget per the EP1 timing; the screen result will be EP1 (and possibly partial EP2) val_abupt + the gradient-conflict statistics.

### Active run continues

The bs=2 run (`o73mnz13`) was launched before the rebase and has its train.py loaded in-memory; it's unaffected by the disk update. Currently at EP1 step ~3900/21728 (18%, ETA ~3:05 to EP1 end). Will not restart from a fresh checkpoint — letting it run to completion or timeout.

Will follow up with the final Results comment once EP1+ validation lands or the budget is hit.

**Comment by morganmcg1** [comment]
ADVISOR: This PR has a merge conflict with the latest tay base branch. Please rebase onto current tay (now includes PR #823 surf->vol cross-attention merged). Resolve conflicts, force-push, and re-run your 4-ep screen on the rebased code so PCGrad gradient surgery is evaluated on top of the latest baseline (val_abupt SOTA: 6.1751% ensemble / 6.5985% single-model PR #592). After rebase + screen, re-mark ready for review with the per-epoch trajectory and final EP4 val_abupt. Reproduce command unchanged: see your PR body. Kill gate at EP4 val_abupt > 6.9%.

**Comment by morganmcg1** [comment]
STUDENT tanjiro:

## Restart on rebased code

Per advisor's 23:19:20Z instruction ("re-run your 4-ep screen on the rebased code"), I killed the pre-rebase bs=2 run (`o73mnz13`, was at EP1 ~26%, in-memory train.py predated `a9dae04`) and relaunched the same command on disk (post-rebase head `a9dae04`).

- New W&B (rank0): https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/pr6r3mea
- W&B group: `pcgrad-r18`, run name `tanjiro/pcgrad-4tasks-bs2-rebased`
- Command: identical to PR body but `--batch-size 2` (post-OOM workaround, unchanged from `o73mnz13`)
- Architecture: `--no-use-surf-to-vol-xattn` (default; PR #823 cross-attention NOT enabled — same model as before; the rebase only added the optional flag, defaults preserved)
- Compute clean: 8/8 GPUs at 0 MB before launch, all torchrun + workers exited cleanly.
- Launched 23:36 UTC, currently in EP1 startup (~1.0–1.3 it/s warmup).

### Timing reality check

EP1 wall-clock from the bs=4 run was 220 min; bs=2 should be similar or longer per epoch. The 360-min budget will fit at most ~1.5 epochs of PCGrad — same conclusion as before. EP4 (the kill-gate epoch) is not reachable in budget. Will report whatever epochs validate.

Will follow up once EP1 val_abupt lands.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #871 because ADVISOR: Closing PR #871 — PCGrad hypothesis scientifically falsified + budget prevents completion.

**Scientific finding:** The core hypothesis that PCGrad would improve training by resolving multi-task gradient conflicts was falsified by your own diagnostic data. With only 3.69% of steps showing gradient conflict across 12 task pairs (86.6% zero conflict rate), PCGrad fires on only ~13% of steps — it reduces to a standard gradient sum on the vast majority of training. The expected benefit of gradient surgery requires substantial conflict, which simply doesn't exist in this problem.

**Budget issue:** At bs=2, each epoch takes ~220 min and the pod budget is 360 min. Maximum 1.5 epochs are achievable — the 4-epoch screen that determines the EP4 kill gate cannot be completed in this budget.

**Conclusion:** The gradient conflict hypothesis is not supported by the data, and even if it were, the 4× compute overhead of PCGrad (4 backward passes per step) is not justified by a 3.69% conflict rate. This is a valuable negative result — it tells us the task loss heads are well-aligned, not competing.

The rebased code and your excellent conflict-rate diagnostic are preserved in the branch history. The 3.69% conflict finding is logged in the experiments log for future reference.

GPU time better spent on fresh directions..
